### ISSUE TITLE
Initial data channel state is inconsistent

### ISSUE THREAD

**Issue by nils-ohlmeier** [w3c/webrtc-pc#1761]
In section [6.2](https://w3c.github.io/webrtc-pc/#rtcdatachannel) of the spec we have the following sentence:  
"An RTCDataChannel, created with createDataChannel or dispatched via an RTCDataChannelEvent, MUST initially be in the connecting state."

But then in the [RTCDataChannelState Enumeration state](https://w3c.github.io/webrtc-pc/#dom-rtcdatachannelstate) for 'open' it says:
"This is the initial state of an RTCDataChannel object dispatched as a part of an RTCDataChannelEvent."

I'm guessing that these two sentences were written before the "underlying data transport" concept got introduced.

To me it would make the most sense if these get changed to describe that 'connecting' is the initial state if the underlying data transport is not yet connected, and 'open' should be state if the underlying data transport is connected.

**Comment by lgrahl** [comment]
Upvoting your suggestion.

(In hindsight, I think it would have created a less clunky API if creating a data channel would have been a promise that resolves once that channel is open. This way we could have avoided the quirky *temporary stream id* thing. But it would break API... sadly...)

**Comment by nils-ohlmeier** [comment]
After talking to @jan-ivar to this is it turns out that my suggestion is not a good idea from an API's perspective.

In the _RTCDataChannelEvent_ event handler is the first time you get a reference to a newly created channel. Thus this is also the first time you can install an _onopen_ callback on the new channel. But in the scenario where the underlying transports are connected already if we then start any new channel in _open_ right away, every user of the API would have to have an if condition in the _ondatachannel_ event handler distinguishing between the first and all subsequent channels.

So for the purpose of having a consistent API which always behaves the same it seems a new channel in the _RTCDataChannelEvent_ should always start in the 'connecting' state, so that you can always setup an _onopen_ callback, which executes always the same code no matter what the initial state of the underlying transports is/was.

But that means that we would have to remove the concept of 'underlying transport' from a couple of place in the data channel sections.

**Comment by taylor-b** [comment]
> So for the purpose of having a consistent API which always behaves the same it seems a new channel in the RTCDataChannelEvent should always start in the 'connecting' state, so that you can always setup an onopen callback, which executes always the same code no matter what the initial state of the underlying transports is/was.

I suspected this was the reasoning; makes enough sense to me.

**Comment by jan-ivar** [comment]
> This way we could have avoided the connecting state and the quirky temporary stream id thing

FWIW you can avoid `ondatachannel` and the `"connecting"` logic using [negotiated](https://jsfiddle.net/jib1/ovbc1uzf/) data channels.

**Comment by lgrahl** [comment]
We do have a slight problem now: After `ondatachannel` has been fired, the channel is in a limbo state. It is technically open and nothing prevents us from sending anything **but** `.send` throws in case the channel is not open. Requiring to register `.onopen` and force waiting for that to fire is awkward.

Would it be possible to say the following:
Set `ReadyState` to `open` before firing `ondatachannel` and fire `onopen` afterwards.

**Comment by taylor-b** [comment]
Actually: why not just say "`createDataChannel` always starts as `connecting`, `ondatachannel` always starts as `open`"?

**Comment by lgrahl** [comment]
Because what @nils-ohlmeier mentioned: https://github.com/w3c/webrtc-pc/issues/1761#issuecomment-364297317

**Comment by taylor-b** [comment]
The downside brought up was:

> every user of the API would have to have an if condition in the ondatachannel event handler distinguishing between the first and all subsequent channels

But `datachannel` fires in response to receiving a "DATA_CHANNEL_OPEN" message, meaning the SCTP association is guaranteed to be established, and even the first channel could start as `open`.

I guess your proposal is better though; if you can send immediately _and_ get an event, that should make everyone happy.

**Comment by aboba** [comment]
@lgrahl Can you prepare a PR?

**Comment by adam-be** [comment]
> "This is the initial state of an RTCDataChannel object dispatched as a part of an RTCDataChannelEvent."

The sentence above is from the time when channels delivered by the 'datachannel' event were always open. I guess it should have been removed when we moved to the always 'connecting' approach. As long as we "delay" the 'datachannel' event enough so we consistently can say that all channels dispatched with the event always are open, I like that approach better. At least now I can't see any major drawbacks with the hybrid solution suggested here.
